File: app/src/main/java/cat/teknos/pokedex/adapter/PokemonListAdapter.java
package cat.teknos.pokedex.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.google.gson.internal.LinkedTreeMap;

import java.util.ArrayList;
import java.util.Objects;

import cat.teknos.pokedex.R;
import cat.teknos.pokedex.api.Pokemon;

public class PokemonListAdapter extends RecyclerView.Adapter<PokemonListAdapter.ViewHolder> {

    private static Context context;
    private final ArrayList<Pokemon> dataset;

    public PokemonListAdapter(Context context){
        PokemonListAdapter.context = context;
        dataset = new ArrayList<>();

    }

    public void addPokemon(ArrayList<Pokemon> listaPokemon) {
        dataset.addAll(listaPokemon);
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_pokemon, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        Object p = dataset.get(position);
        LinkedTreeMap<Pokemon, Object> linkedTreeMap = (LinkedTreeMap)p;

        holder.tvPokemonName.setText(Objects.requireNonNull(linkedTreeMap.get("name")).toString());
        String numeroStr = Objects.requireNonNull(linkedTreeMap.get("url")).toString();
        String[] NumeroPokemon = numeroStr.split("/");
        String numeroFinalStr = NumeroPokemon[6];
        System.out.println(numeroFinalStr);

        Glide.with(context)
                .load( "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/"
                        + numeroFinalStr + ".png")
                .centerCrop()
                .diskCacheStrategy(DiskCacheStrategy.ALL)
                .into(holder.ivPokemon);
    }

    @Override
    public int getItemCount() {
        return dataset.size();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder{
        private final ImageView ivPokemon;
        private final TextView tvPokemonName;

        public ViewHolder(View itemView){
            super(itemView);

            ivPokemon = itemView.findViewById(R.id.ivPokemon);
            tvPokemonName = itemView.findViewById(R.id.tvPokemonName);

            ivPokemon.setOnClickListener(v -> {
                Toast.makeText(context, tvPokemonName.getText().toString().toUpperCase(), Toast.LENGTH_SHORT).show();
            });
        }
    }
}

File: app/src/main/java/net/atos/pokedex/adapter/PokemonListAdapter.kt
package net.atos.pokedex.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.cardview.widget.CardView
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.google.gson.internal.LinkedTreeMap
import net.atos.pokedex.R
import net.atos.pokedex.api.Pokemon
import java.util.Locale
import java.util.Objects

class PokemonListAdapter(context: Context?) : RecyclerView.Adapter<PokemonListAdapter.ViewHolder>() {

    private val dataset: ArrayList<Pokemon>

    init {
        Companion.context = context
        dataset = ArrayList()
    }

    fun addPokemon(pokemonList: ArrayList<Pokemon>?) {
        dataset.addAll(pokemonList!!)
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_pokemon, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val p: Any = dataset[position]
        val linkedTreeMap: LinkedTreeMap<*, *> = p as LinkedTreeMap<*, *>
        holder.tvPokemonName.text =
            Objects.requireNonNull(linkedTreeMap["name"]).toString()
        val strNumber = Objects.requireNonNull(linkedTreeMap["url"]).toString()
        val pokemonNumber = strNumber.split("/".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        val finalStrNumber = pokemonNumber[6]
        println(finalStrNumber)
        Glide.with(context!!)
            .load(
                "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/"
                        + finalStrNumber + ".png"
            )
            .centerCrop()
            .diskCacheStrategy(DiskCacheStrategy.ALL)
            .into(holder.ivPokemon)
    }

    override fun getItemCount(): Int {
        return dataset.size
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val cardPokemon: CardView = itemView.findViewById(R.id.pokemon_item)
        val ivPokemon: ImageView = itemView.findViewById(R.id.ivPokemon)
        val tvPokemonName: TextView = itemView.findViewById(R.id.tvPokemonName)

        init {
            cardPokemon.setOnClickListener {
                val pokemonName = tvPokemonName.text.toString().uppercase(Locale.getDefault())
                val toastContext = itemView.context.applicationContext
                Toast.makeText(toastContext, pokemonName, Toast.LENGTH_SHORT).show()
            }
        }
    }

    companion object {
        private var context: Context? = null
    }
}